include/materials/ADComputeISoilStress.h

```diff
@@ -0,0 +1,162 @@
+/*************************************************/
+
+// This code was implemented in colloboration with Ozgun Numanoglu
+// (numanog2@illinois.edu) and Omar Baltaji (baltaji2@illinois.edu) from UIUC.
+
+/**
+ * This material class calculates the total stress for the  nonlinear hysteretic
+ * soil model I-Soil.
+ **/
+#pragma once
+
+#include "ADComputeFiniteStrainElasticStress.h"
+
+#define usingComputeISoilStressMembers                                                             \
+  usingComputeFiniteStrainElasticStressMembers;                                                    \
+  using ADComputeISoilStress<compute_stage>::computeStress;                                        \
+  using ADComputeISoilStress<compute_stage>::computeSoilLayerProperties;
+
+template <ComputeStage>
+class ADComputeISoilStress;
+
+declareADValidParams(ADComputeISoilStress);
+
+template <ComputeStage compute_stage>
+class ADComputeISoilStress : public ADComputeFiniteStrainElasticStress<compute_stage>
+{
+public:
+  ADComputeISoilStress(const InputParameters & parameters);
+
+  //  virtual void initialSetup() override;
+
+protected:
+  virtual void computeQpStress() override;
+
+  /// Calculates the total stress as a sum of the stresses from the individual
+  /// elastic-perfectly plastic stress strain curves.
+  virtual void computeStress();
+
+  virtual void initQpStatefulProperties() override;
+
+  /// Deconstructs the backbone curve into a set of elastic-perfectly-plastic
+  /// curves, each comprising a Young's modulus and yield stress.
+  void computeSoilLayerProperties(std::vector<std::vector<ADReal>> & youngs,
+                                  std::vector<std::vector<ADReal>> & yield_stress,
+                                  const std::vector<std::vector<Real>> & backbone_stress,
+                                  const std::vector<std::vector<Real>> & backbone_strain,
+                                  const std::vector<unsigned int> & layer_ids,
+                                  const std::vector<Real> & poissons_ratio,
+                                  const std::string & name);
+
+  /// Base names given to each elastic perfectly plastic curve.
+  std::vector<std::string> _base_models;
+
+  /// Material property which stores the current stress for each elastic
+  /// perfectly plastic curve.
+  std::vector<ADMaterialProperty(RankTwoTensor) *> _stress_model;
+
+  /// Material property which stores the old stress for each elastic perfectly
+  /// plastic curve.
+  std::vector<const MaterialProperty<RankTwoTensor> *> _stress_model_old;
+
+  /// Yield stress for each elastic perfectly plastic curve for each soil layer.
+  std::vector<std::vector<ADReal>> _yield_stress;
+
+  /// Young's modulus for each elastic perfectly plastic curve for each soil
+  /// layer.
+  std::vector<std::vector<ADReal>> _youngs;
+
+  /// The coupled variable field providing the soil layer id.
+  const VariableValue & _soil_layer_variable;
```

Comment:
I guess this doesn't have to be an `ADVariableValue`?

Reply:
This is a constant value which is used as the index to a material property vector and it is only used to determine the appropriate the material properties corresponding to each element. This variable is not used in the residual calculation.

Reply:
👍